File: frontend/scripts/check-styles.py
#!/usr/bin/env python3
"""Fail if a class the components render has no rule in any stylesheet.

Neither `tsc` nor the bundler can catch this: `className="dialog"` is just a
string, so a stylesheet that lost half its rules still typechecks and still
builds. This closes that gap -- it is the one check that would have caught a
truncated `styles.css` immediately rather than after a visual read of the page.

Only plain literal class names are checked. A name assembled at runtime
(`` `row is-${state}` ``) is matched per-word where the words are literal, and
interpolated fragments are skipped rather than guessed at.
"""

from __future__ import annotations

import re
import sys
from pathlib import Path

SRC = Path(__file__).resolve().parent.parent / "src"

# A class name we accept as literal: lowercase, digits and dashes. This
# deliberately excludes anything with `${`, quotes or capitals, so template
# fragments cannot be mistaken for names. A trailing dash means the name was
# cut off by an interpolation (`is-${state}`), so it is a prefix, not a name.
NAME = re.compile(r"^[a-z][a-z0-9-]*[a-z0-9]$|^[a-z]$")

# `className="a b"`, `className={`a b`}` and `className={"a b"}`. The template
# form keeps whatever literal words sit outside the `${...}` holes.
CLASS_ATTR = re.compile(r'className=(?:"([^"]*)"|\{`([^`]*)`\}|\{"([^"]*)"\})')

HOLE = re.compile(r"\$\{[^}]*\}")

# Names that live only in markup we do not own, or are set on <html>/<body>
# from JS rather than rendered by a component.
EXEMPT: set[str] = set()


def defined_classes() -> set[str]:
    css = "".join(path.read_text() for path in sorted(SRC.glob("*.css")))
    return set(re.findall(r"\.([A-Za-z][A-Za-z0-9_-]*)", css))


def rendered_classes() -> dict[str, str]:
    """Every literal class name, mapped to the file that renders it."""
    found: dict[str, str] = {}
    for path in sorted(SRC.rglob("*.tsx")):
        for groups in CLASS_ATTR.findall(path.read_text()):
            # Drop the `${...}` holes first. Their contents are expressions, and
            # an identifier inside one (`mode === value`) is not a class name.
            text = HOLE.sub(" ", " ".join(groups))
            for word in text.split():
                if NAME.match(word):
                    found.setdefault(word, str(path.relative_to(SRC)))
    return found


def main() -> int:
    defined = defined_classes()
    rendered = rendered_classes()
    missing = sorted(name for name in rendered if name not in defined | EXEMPT)

    if missing:
        print(f"{len(missing)} rendered class(es) have no CSS rule:\n")
        for name in missing:
            print(f"  .{name:<24} rendered by src/{rendered[name]}")
        print("\nEither add a rule or stop rendering the class.")
        return 1

    print(f"styles ok: {len(rendered)} rendered classes all have rules")
    return 0


if __name__ == "__main__":
    sys.exit(main())

File: frontend/scripts/check-styles.mjs
#!/usr/bin/env node
/**
 * Fail if a class the components render has no rule in any stylesheet.
 *
 * Neither `tsc` nor the bundler can catch this: `className="dialog"` is just a
 * string, so a stylesheet that lost half its rules still typechecks and still
 * builds. This closes that gap -- it is the one check that would have caught a
 * truncated `styles.css` immediately rather than after a visual read of the
 * page.
 *
 * Written in Node rather than Python because it runs inside `npm run build`,
 * and that script also runs in the Docker web stage, which is `node:alpine`
 * and has no Python. A guard that only works on the author's machine is not a
 * guard.
 *
 * Only plain literal class names are checked. A name assembled at runtime is
 * matched per-word where the words are literal; interpolated fragments are
 * skipped rather than guessed at.
 */

import { readdirSync, readFileSync, statSync } from "node:fs";
import { dirname, join, relative } from "node:path";
import { fileURLToPath } from "node:url";

const SRC = join(dirname(fileURLToPath(import.meta.url)), "..", "src");

// A class name we accept as literal: lowercase, digits and dashes. This
// deliberately excludes anything with a hole, quotes or capitals, so template
// fragments cannot be mistaken for names. A trailing dash means the name was
// cut off by an interpolation, so it is a prefix, not a name.
const NAME = /^(?:[a-z][a-z0-9-]*[a-z0-9]|[a-z])$/;

// `className="a b"`, a template literal, and `className={"a b"}`. The template
// form keeps whatever literal words sit outside the holes.
const CLASS_ATTR = /className=(?:"([^"]*)"|\{`([^`]*)`\}|\{"([^"]*)"\})/g;

const HOLE = /\$\{[^}]*\}/g;

// Names that live only in markup we do not own, or are set on <html>/<body>
// from JS rather than rendered by a component.
const EXEMPT = new Set();

function walk(dir, suffix, out = []) {
  for (const entry of readdirSync(dir).sort()) {
    const full = join(dir, entry);
    if (statSync(full).isDirectory()) walk(full, suffix, out);
    else if (entry.endsWith(suffix)) out.push(full);
  }
  return out;
}

function definedClasses() {
  const css = walk(SRC, ".css")
    .filter((path) => dirname(path) === SRC)
    .map((path) => readFileSync(path, "utf8"))
    .join("");
  const found = new Set();
  for (const match of css.matchAll(/\.([A-Za-z][A-Za-z0-9_-]*)/g)) found.add(match[1]);
  return found;
}

/** Every literal class name, mapped to the file that renders it. */
function renderedClasses() {
  const found = new Map();
  for (const path of walk(SRC, ".tsx")) {
    for (const match of readFileSync(path, "utf8").matchAll(CLASS_ATTR)) {
      // Drop the holes first. Their contents are expressions, and an identifier
      // inside one is not a class name.
      const text = [match[1], match[2], match[3]]
        .filter(Boolean)
        .join(" ")
        .replace(HOLE, " ");
      for (const word of text.split(/\s+/)) {
        if (NAME.test(word) && !found.has(word)) found.set(word, relative(SRC, path));
      }
    }
  }
  return found;
}

const defined = definedClasses();
const rendered = renderedClasses();
const missing = [...rendered.keys()]
  .filter((name) => !defined.has(name) && !EXEMPT.has(name))
  .sort();

if (missing.length > 0) {
  console.log(`${missing.length} rendered class(es) have no CSS rule:\n`);
  for (const name of missing) {
    console.log(`  .${name.padEnd(24)} rendered by src/${rendered.get(name)}`);
  }
  console.log("\nEither add a rule or stop rendering the class.");
  process.exit(1);
}

console.log(`styles ok: ${rendered.size} rendered classes all have rules`);
